[http] add llhttp parser implementation

asraa opened the pull request:
Signed-off-by: Asra Ali <asraa@google.com>

Commit Message: Add llhttp parser implementation.
Risk Level: High, but off by default.
Testing: H/1 codec impl and unit tests that use actual codecs have two targets now, a legacy and new version. Integration tests are parametrized by enabling the runtime flag via the config setting and is enabled in compile_time_options. Codec fuzz test enables the new parser.
Release Notes: Added
Runtime guard: `envoy.reloadable_features.enable_new_http1_parser`. Off by default.
Fixes: https://github.com/envoyproxy/envoy/issues/5155
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=25777 (see note below)

**CHANGES**
* We use llhttp to detect and reject invalid header characters. Thus, `http1.invalid_characters` response codes appear as `http1.codec_error`. and the status message reads `HPE_INVALID_HEADER_TOKEN`. 
* It also appears that llhttp handles invalid methods errors after message begin, whereas http-parser handles it at message begin. I only noticed this when it triggered a codec fuzzer failure when a second H/1 stream started. http-parser catches the invalid method failure and exits early, whereas llhttp hits the latent bug: there were no stream resets when responses were completed before requests are finished. In the normal case, we expect higher level code to reset the stream. 

**NOTES FOR CLEANUP**
* Our invalid character check in the codec can be removed when the legacy code is removed.
* The secondary impl layer can be removed when the legacy code is removed. Actually, the parser interface should totally be removed since there's a slight performance dip from using it.

moderation commented on bazel/repository_locations.bzl:
```diff
@@ -490,6 +490,18 @@ REPOSITORY_LOCATIONS_SPEC = dict(
+        project_name = "llhttp",
+        project_desc = "Parser for HTTP messages written in C",
+        project_url = "https://github.com/nodejs/llhttp",
+        version = "3.0.1",
+        sha256 = "6ae898e5c5e8081be68d4d93d24349c8a4f4d0c7e3eb89f0e35db5a7f15d59fa",
+        strip_prefix = "llhttp-release-v{version}",
+        urls = ["https://github.com/nodejs/llhttp/archive/release/v{version}.tar.gz"],
+        use_category = ["dataplane_core"],
+        release_date = "2020-02-16",
```
Wondering if we should use the more current `4.0.0` release - https://github.com/nodejs/llhttp/releases/tag/release%2Fv4.0.0 ? `3.0.1` is marked as the latest release however. I'll try and ping @indutny to get guidance on what we should target

indutny replied:
Sorry for confusion. 4.0.0 is the latest version, I just had to do an extra backport release in 3.x branch.

asraa replied:
thank you both for that! will update.

moderation commented:
/lgtm deps

asraa commented:
Very confused about the ASan error. It is reproducible via bazel test, but not by gdb or running the binary directly.

When I bazel test, it appears the function doesn't enter `llhttp_execute` at all but somehow returns from it.

antoniovicente commented on source/common/http/http1/parser_impl.h:
```diff
@@ -0,0 +1,39 @@
+namespace Http {
+namespace Http1 {
+
+class HttpParserImpl : public Parser {
```
Small tests for HttpParserImpl?

asraa replied:
minimal tests added with a no-op parser callback.

antoniovicente commented on source/common/http/http1/parser_impl.h:
```diff
@@ -0,0 +1,39 @@
+
+private:
+  class Impl;
+  std::unique_ptr<Impl> impl_;
```
Why the impl_ and extra level of unique_ptrs in the parser implementation?

asraa replied:
Leftover from @derekargueta -- do you know why? Maybe something went wrong in construction/destruction?

derekargueta replied:
When trying to get http_parser and llhttp to play nicely together for the transition period where we'll have to support both libraries at runtime, I was running into the issue of colliding symbols between the two libraries, e.g. `#define`'d `HTTP_ERRNO_MAP`, `HTTP_METHOD_MAP`, etc.

So I resorted to using the PImpl pattern to contain the libraries within separate compilation units (transitive inclusion would still hit this issue, e.g. in the factory). Another solution could be patching one of the libraries to change the defined names, but that seemed pretty intrusive. Once http-parser is removed, we wouldn't have this issue.

This was one reason I initially started this with toggling the library via a compile flag, but on code review it was expressed that we want the parser library to be runtime configurable.

antoniovicente replied:
Would it be possible to add this explanation to the header file and possibly a TODO to remove the Impl once it is possible to do so?

asraa replied:
Done.

antoniovicente commented on test/integration/base_integration_test.cc:
```diff
@@ -54,6 +54,9 @@ BaseIntegrationTest::BaseIntegrationTest(const InstanceConstSharedPtrFn& upstrea
       default_log_level_(TestEnvironment::getOptions().logLevel()) {
+#ifdef ENVOY_USE_NEW_HTTP1_PARSER_IN_INTEGRATION_TESTS
+  config_helper_.addRuntimeOverride("envoy.reloadable_features.enable_new_http1_parser", "true");
+#endif
```
I take it that all existing tests pass with the new parser?  Are there any known differences between the two parsers?

asraa replied:
I added some description, there shouldn't be any difference from user perspective except the one relating to the lenient header comment.

antoniovicente commented on bazel/repository_locations.bzl:
```diff
@@ -490,6 +490,18 @@ REPOSITORY_LOCATIONS_SPEC = dict(
+        use_category = ["dataplane_core"],
+        release_date = "2021-04-03",
+        cpe = "N/A",
+    ),
```
the CI dependencies precheck is failing, please take a look

asraa replied:
The release date differing from GitHub UI will always fail me.

moderation replied:
Hey @asraa I'm going to fix this shortly. The current code looks at the UTC for the last commit for the release. This can be different from when the release was actually cut. With my new version you'll still have to enter the UTC for the release but at least you don't have to reverse engineer back through commits /cc @htuch

antoniovicente reviewed (commented):
It is really exciting to see progress on Envoy moving to a http1 parser that is under active development.

asraa commented on test/integration/base_integration_test.cc:
```diff
@@ -54,6 +54,9 @@ BaseIntegrationTest::BaseIntegrationTest(const InstanceConstSharedPtrFn& upstrea
       config_helper_(version, *api_, config),
       default_log_level_(TestEnvironment::getOptions().logLevel()) {
+#ifdef ENVOY_USE_NEW_HTTP1_PARSER_IN_INTEGRATION_TESTS
+  config_helper_.addRuntimeOverride("envoy.reloadable_features.enable_new_http1_parser", "true");
```
Codec impl test has two build modes now (legacy_codec_impl_test and codec_impl_test). I could add manual friends (conn pool comes to mind)

I'm not a fan of this solution in all honest and would prefer a define directly in the parser code so that I could just flip on the new parser at build time with a define, but it doesn't seem to typically be the pattern...

asraa commented on source/common/http/http1/parser_impl.cc:
```diff
@@ -0,0 +1,226 @@
+  void resume() { llhttp_resume(&parser_); }
+
+  ParserStatus pause() {
+    // llhttp can only pause by returning a paused status in user callbacks.
```
The http_parser implementation of `pause()` uses the http parser method to pause (https://github.com/nodejs/http-parser/blob/ec8b5ee63f0e51191ea43bb0c6eac7bfbff3141d/http_parser.h#L438) and an OK. That's the same as returning pause from llhttp parser. I think it's already reconciled

antoniovicente replied:
I guess this is an area to cleanup once llhttp is the only implementation.

In the current world I would consider something like:
ParserStatus maybePause(ParserStatus status);

The http_parser version would detect status == Paused and call its pause method and return Success, while the llhttp version would just return the status that was passed in.

asraa replied:
> In the current world I would consider something like:

To me that seems strange because Paused is the only status that maybePause would ever call. 

Technically returning HPE_PAUSED from http-parser will also cause the parser to pause, but I don't trust making the change because there's very little http-parser docs and I don't know if it changes how it handles errors. (All http_parser_pause() does is cause HPE_PAUSED to be returning from execute. we could just return it directly).

asraa commented on source/common/http/http1/parser_impl.cc:
```diff
@@ -0,0 +1,226 @@
+    return ParserStatus::NoBody;
+  case 2:
+    return ParserStatus::NoBodyData;
+  case 21:
```
Only HPE_PAUSED, unfortunately.

asraa commented on source/common/http/http1/codec_impl.cc:
```diff
@@ -766,6 +771,7 @@ StatusOr<ParserStatus> ConnectionImpl::onHeadersComplete() {
       return codecProtocolError("http/1.1 protocol error: unsupported transfer encoding");
     }
   }
+  parser_->hasContentLength(request_or_response_headers.ContentLength() != nullptr);
```
https://github.com/nodejs/llhttp/blob/aa07189d39b8d0da9d5a887d93761014a45804bf/test/request/content-length.md#error-on-simultaneous-content-length-and-transfer-encoding-identity

It's handled with an error. Actually this is just because llhttp doesn't have an initial value for unset content length. `http_parser`  uses all 1 bits set to indicate there is no content length. Added comment.

antoniovicente replied:
We should have tests in envoy to cover these cases.  This is related to my question regarding use of lenient parsing further down in this file.  It seems that those lenient options were added for the benefit of Envoy's H1 implementation.

asraa replied:
https://github.com/envoyproxy/envoy/blob/56d3f8ab41cefbb6ca485f882f47b7e1703a3fa7/test/common/http/http1/codec_impl_test.cc#L1945 

There are tests for both content length and chunked.

indutny commented:
I would appreciate if you could upgrade llhttp to 5.1.0 or 4.1.0, which was just released: https://github.com/nodejs/llhttp/pull/95

antoniovicente commented on bazel/repository_locations.bzl:
```diff
@@ -494,12 +494,12 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         project_url = "https://github.com/nodejs/llhttp",
-        version = "4.0.0",
-        sha256 = "0b83fd2db4300e85b3a73763b4c2599998089722c2ab8507383e02517b274cb7",
+        version = "5.0.0",
```
Apparently there was a 5.1.0 release.

asraa commented on source/common/http/http1/parser_impl.cc:
```diff
@@ -0,0 +1,226 @@
+  Impl(llhttp_type_t type) {
+    llhttp_init(&parser_, type, &settings_);
+    llhttp_set_lenient_chunked_length(&parser_, 1);
+    llhttp_set_lenient_headers(&parser_, 1);
```
Well, maybe. We have our own code for header validation. It would change error code/messages. But since this feature has to be turned on and deprecated maybe this is a great time to do this and remove some codec code.

Will do. And leaving comment in codec code that we will be able to remove this after removing llhttp.

asraa commented on source/common/http/http1/parser_impl.cc:
```diff
@@ -0,0 +1,226 @@
+public:
+  Impl(llhttp_type_t type) {
+    llhttp_init(&parser_, type, &settings_);
+    llhttp_set_lenient_chunked_length(&parser_, 1);
```
I could pipe https://github.com/envoyproxy/envoy/blob/c93d486c2d303dbf2cc9f88717de0bfd66e1afdb/include/envoy/http/codec.h#L438 in to the parser. Like the lenient header validation, it would change the response code/message. WDYT?

antoniovicente replied:
What code in Envoy is generating the error response when both transfer-encoding: chunked and content-length is set?

asraa replied:
https://github.com/envoyproxy/envoy/blob/c93d486c2d303dbf2cc9f88717de0bfd66e1afdb/source/common/http/http1/codec_impl.cc#L750-L753

wrowe replied:
Line 718 already circumvents this test.

wrowe replied:
The test at 747 looks altogether incorrect, whether chunked is disallowed doesn't impact the rule of 'never both'

asraa replied:
Why does 718 circumvent this block?

`codec_settings_.allow_chunked_length_`: allows Envoy to process requests with both C-L and TE without rejecting (and remove C-L). So I think it does impact the rule of never both..

wrowe replied:
So, it should be read as allow_chunked_and_content_length_ ... that is sensible (very unwise, but not a default nor a defect)

wrowe replied:
And the same toggle check should probably occur at 720.

asraa replied:
I'm not sure I understand why. That's removing a meaningless 0 content-length and otherwise rejecting the request since we shouldn't have a payload for CONNECT, not because TE: chunked is present.

antoniovicente replied:
It seems to me that the integration tests pass with the current code.  I'm curious about which fail if we were to flip the value of these llhttp library settings.

asraa replied:
So long as we flipped based on the codec setting, the only change would be the status messages/response details:
```
Expected equality of these values:
  status.message()
    Which is: "http/1.1 protocol error: HPE_UNEXPECTED_CONTENT_LENGTH"
  "http/1.1 protocol error: both 'Content-Length' and 'Transfer-Encoding' are set."
    Which is: 0x3a24fa
```

I think adding a comment that our codec code handling this could be removed after clean up would make sense.

asraa commented:
Updated, sorry for the delay! (This is an evening thing right now)

* For unit test dupes: only mixed_conn_pool test used an actual parser (I added an ASSERT(false) in the default parser and ran all the http/... conn_pool/... to grab failures).
* A latent codec fuzz bug needed fixing, I noted why it came up in the PR description. 
* llhttp version's bumped.
* Let me know what you think about the lenient chunked length setting. I switched over to llhttp's strict header validation since Envoy no longer provides the option to disable that.

asraa commented:
Still working on the asan issue (comment above) and a tsan one. TSan is from running existing parser, reproducible, so something must have went wrong there...

antoniovicente commented on test/common/http/codec_impl_corpus/clusterfuzz-testcase-minimized-codec_impl_fuzz_test-5689430000795648:
```diff
@@ -0,0 +1,62 @@
+actions {
```
What fixes are needed to make this corpus entry run successfully?  I assume it's the changes to the fuzz harness?

asraa replied:
It does run successfully now, with the changes in the diff that close that reset stream when the response ends before the request does.

antoniovicente commented on test/common/http/codec_impl_fuzz_test.cc:
```diff
@@ -681,6 +708,11 @@ DEFINE_PROTO_FUZZER(const test::common::http::CodecImplFuzzTestCase& input) {
     TestUtility::validate(input);
     codecFuzz(input, HttpVersion::Http1);
     codecFuzz(input, HttpVersion::Http2);
+    // Enable legacy HTTP1 codec.
```
This comment seems out of sync with what is done below.  I recommend you explicitly set the runtime flag for the 2 HTTP1 runs, do not rely on the default value of the runtime flag.

asraa replied:
Oops the comment was incorrect. Line 709 is legacy, this is new. I will explicitly add the runtime vals.

antoniovicente commented:
Still seeing ASAN errors.

/wait

moderation commented on bazel/repository_locations.bzl:
```diff
@@ -490,6 +490,18 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         release_date = "2020-07-10",
         cpe = "cpe:2.3:a:nodejs:node.js:*",
     ),
+    com_github_nodejs_llhttp = dict(
+        project_name = "llhttp",
+        project_desc = "Parser for HTTP messages written in C",
+        project_url = "https://github.com/nodejs/llhttp",
+        version = "5.1.0",
+        sha256 = "7140bbafa30f9cad4d20a0065953bd19bcc11203ab9a901ce7dea1b8182c8c78",
+        strip_prefix = "llhttp-release-v{version}",
+        urls = ["https://github.com/nodejs/llhttp/archive/release/v{version}.tar.gz"],
+        use_category = ["dataplane_core"],
+        release_date = "2021-04-06",
+        cpe = "cpe:2.3:a:nodejs:node.js:*",
+    ),
```
There is another version bump - https://github.com/nodejs/llhttp/releases/tag/release%2Fv6.0.0

indutny replied:
Yeah, sorry for that. I've merged pull request for RTSP support without realizing that the exported RTSP methods would interact badly with existing Node.js ecosystem. Had to do a major release because the C headers changed a bit. I don't think there's anything of concern for envoyproxy, but glad that you keep a note of a new release!

asraa commented:
Yeah :/ still working on it. The error only appears in `bazel test` and not when running the binary so debugging is pretty confusing to me.

indutny commented:
Can I help?

asraa commented:
Perhaps!! So when we dispatch data (particularly in a test tested that large trailers are rejected, see the buffer constructed [here](https://github.com/envoyproxy/envoy/blob/38f673821df18513fb9c35edb0b45f59e778ad64/test/common/http/http1/codec_impl_test.cc#L243-L253) with a trailer of ["big: " + std::string(60*1024, 'q')](https://github.com/envoyproxy/envoy/blob/38f673821df18513fb9c35edb0b45f59e778ad64/test/common/http/http1/codec_impl_test.cc#L2797), I'm getting this strange stack trace:

```
source/common/http/http1/parser_impl.cc:107:9: runtime error: load of value 4294967295, which is not a valid value for type 'llhttp_errno_t' (aka 'llhttp_errno')
    #0 0xfaad5a5 in Envoy::Http::Http1::HttpParserImpl::Impl::execute(char const*, int) /proc/self/cwd/source/common/http/http1/parser_impl.cc:107:9
    #1 0xfaab614 in Envoy::Http::Http1::HttpParserImpl::execute(char const*, int) /proc/self/cwd/source/common/http/http1/parser_impl.cc:177:17
    #2 0xfa324b9 in Envoy::Http::Http1::ConnectionImpl::dispatchSlice(char const*, unsigned long) /proc/self/cwd/source/common/http/http1/codec_impl.cc:619:31
    #3 0xfa2e188 in Envoy::Http::Http1::ConnectionImpl::dispatch(Envoy::Buffer::Instance&) /proc/self/cwd/source/common/http/http1/codec_impl.cc:582:30
    #4 0xfa34fea in virtual thunk to Envoy::Http::Http1::ConnectionImpl::dispatch(Envoy::Buffer::Instance&) /proc/self/cwd/source/common/http/http1/codec_impl.cc
    #5 0xd51a921 in Envoy::Http::Http1ServerConnectionImplTest::testTrailersExceedLimit(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool) /proc/self/cwd/test/common/http/http1/codec_impl_test.cc:253:22
```

I know the parser is executing `llhttp_execute` from here: https://github.com/envoyproxy/envoy/blob/53232a4394fc15850fa153a9b4d1ea112930c969/source/common/http/http1/parser_impl.cc#L101
but when I insert logging statements with a patch in `llhttp__internal_execute`, I see absolutely nothing.  

Again, this weirdly only happens in the bazel test environment, and I wish I understood why, since that might tell my something about why the code is not stepping into that function.

indutny commented:
4294967295 is basically -1 for 32bit integers. Did you try testing 32bit build of it?

indutny commented:
Another idea for testing. I think there might be a documentation bug in llhttp api.h. Could you try returning `HPE_USER` instead of -1 [here](https://github.com/envoyproxy/envoy/blob/53232a4394fc15850fa153a9b4d1ea112930c969/source/common/http/http1/parser_impl.cc#L208-L209) ?

asraa commented:
YES! That works. Hah, I had noticed that and brushed it off as maybe what UBSan sets undefined things to by default :P Thank you so much!! Updated.

indutny commented:
@asraa Glad to hear it is working. After looking at llhttp's code I think it is a bug that should be fixed. Perhaps, if you could give me a few days to let me submit a PR and make minor release - it'd be for the best. What do you think?

indutny commented:
Nevermind, I got an idea of a patch that would legitimize return HPE_USER. Thanks for catching this!

indutny commented:
@asraa (Sorry for spurious messages) I know what's going on now. Neither of these [3 callbacks](https://github.com/envoyproxy/envoy/blob/53232a4394fc15850fa153a9b4d1ea112930c969/source/common/http/http1/parser_impl.cc#L49-L64) supports `-1` return value. I'm going to submit a PR and make a minor release of llhttp to add support for that. If you'd like to move forward without being blocked my work - could you change just these three callbacks to return `HPE_USER` instead? The others should not really return `HPE_USER`, only these three.

indutny commented:
[Here's the PR](https://github.com/nodejs/llhttp/pull/104) that adds docs and makes `-1` valid return value. I still recommend going with `HPE_USER`, because you can set your own reason string for it!

asraa commented:
Oh, very weird. So actually http-parser might not have supported it all along either https://github.com/nodejs/http-parser/blob/ec8b5ee63f0e51191ea43bb0c6eac7bfbff3141d/contrib/parsertrace.c#L53-L63 since they always return 0. I think we were always catching errors in subsequent calls. 

> Update: this can't be fixed in lllhttp, because when HPE_USER is returned the user is responsible for setting for setting the error code. From Node.js codebase:

Should we always be setting error strings when returning `HPE_USER/-1`? We return `HPE_USER/-1` wherever there's a codecProtocolError (e.g. [here](https://github.com/envoyproxy/envoy/blob/53232a4394fc15850fa153a9b4d1ea112930c969/source/common/http/http1/codec_impl.cc#L734)) and grab the error back as response [details](https://github.com/envoyproxy/envoy/blob/53232a4394fc15850fa153a9b4d1ea112930c969/source/common/http/conn_manager_impl.cc#L282-L284). When it's an error from the codec, it usually prints something like `HPE_CB_MESSAGE_BEGIN`.

indutny commented:
For `HPE_USER` it is better to set the error reason, but since in the end you're the one who's going to consume it - it is totally up to you! If you return `HPE_USER` without setting the reason - the `llhttp_get_error_reason` function will return `nullptr` or previous error reason.

asraa commented:
Gotcha! Right now we don't use `llhttp_get_error_reason` at all, but after we cleanup the codecs, we could get rid of `codec_error_` entirely and start using this method.

indutny commented:
FWIW, I've just published [llhttp@6.0.1](https://github.com/nodejs/llhttp/releases/tag/release%2Fv6.0.1) that allows returning `-1` from the spans. I still won't recommend going this route, but at least it should not cause asan errors anymore.

antoniovicente commented on source/common/http/http1/parser_impl.cc:
```diff
@@ -0,0 +1,225 @@
+  RcVal execute(const char* slice, int len) {
+    llhttp_errno_t error;
+    if (slice == nullptr || len == 0) {
+      error = llhttp_finish(&parser_);
```
Can we detect the error condition were execute(non-null) is called after execute(null)?

I think this is mostly an issue in the codec interface, end_stream is not plumbed all the way; execute(null) should only happen when buffer is empty and end_stream is true.

asraa replied:
I could add internal state checking?
But we are guaranteed that the len condition will be OK since that's what the caller enforces, and it comes down to whether slice is not nullptr when len > 0.
https://github.com/envoyproxy/envoy/blob/9e263d8d636fb0b52da99a524b906cfc62190ae0/source/common/http/http1/codec_impl.cc#L577

antoniovicente commented on source/common/http/http1/parser_impl.cc:
```diff
@@ -0,0 +1,225 @@
+  case HPE_PAUSED:
+    return ParserStatus::Paused;
+  default:
+    return ParserStatus::Unknown;
```
Coverage report claims that some of these are not covered by tests:
- ParserStatus::Error
- ParserStatus::NoBody
- ParserStatus::NoBodyData
- ParserStatus::Unknown

asraa replied:
The codec only expects success or paused, and otherwise something went wrong (assert checks). I'll change getStatus to isSuccess or isPaused

antoniovicente reviewed (commented):
Looks good, please fix version_history/current.rst after v1.18 branch cut.

asraa commented:
A lot of these coverage issues are probably because the config where llhttp is run in integration tests is not covered by default. The only way to do this would be to dupe every single integration test target. I added the config in the coverage run so it would be easier to check coverage report.

antoniovicente commented on test/run_envoy_bazel_coverage.sh:
```diff
@@ -41,32 +41,40 @@ echo "    VALIDATE_COVERAGE=${VALIDATE_COVERAGE}"
       "--config=fuzz-coverage"
       "--test_tag_filters=-nocoverage")
 else
+  COVERAGE_TARGETS=$COVERAGE_ARGS
```
From CI:

```
2021-04-28T17:34:59.4251043Z Shellcheck: test/run_envoy_bazel_coverage.sh
2021-04-28T17:34:59.5961884Z 
2021-04-28T17:34:59.5963221Z In test/run_envoy_bazel_coverage.sh line 61:
2021-04-28T17:34:59.5963741Z   COVERAGE_TARGETS=$COVERAGE_ARGS
2021-04-28T17:34:59.5964956Z   ^--------------^ SC2178: Variable was used as an array but is now assigned a string.
2021-04-28T17:34:59.5966965Z                    ^------------^ SC2128: Expanding an array without an index only gives the first element.
2021-04-28T17:34:59.5967411Z 
2021-04-28T17:34:59.5967791Z For more information:
2021-04-28T17:34:59.5968516Z   https://www.shellcheck.net/wiki/SC2128 -- Expanding an array without an ind...
2021-04-28T17:34:59.5970085Z   https://www.shellcheck.net/wiki/SC2178 -- Variable was used as an array but...
```

asraa replied:
Yep, but after fixing there's one more problem that running bazel coverage twice doesn't aggregate the coverage profile.

asraa replied:
Should be possible to merge with genhtml after copying them out

asraa commented:
Apparently running bazel coverage twice doesn't aggregate both reports -- i'll try to figure something out to get coverage to show up on CI.

Worst case I can just generate the llhttp coverage report if that's all you're checking for.

asraa commented on source/common/http/http1/parser_impl.cc:
```diff
@@ -0,0 +1,225 @@
+      nread = llhttp_get_error_pos(&parser_) - slice;
+      // Resume after upgrade.
+      if (error == HPE_PAUSED_UPGRADE) {
+        error = HPE_OK;
```
It seems like we handle upgrade by manually setting HPE_PAUSED in the codec (not ever HPE_PAUSED_UPGRADE), so this might be why we don't actually need to follow this contract
https://github.com/nodejs/llhttp/blob/674cf94e975bea73970015eae272088b571cf368/src/native/api.h#L117
@derekargueta do you have more context on this?

asraa replied:
Seems like we could remove it -- the codec handling upgrade is the reason why. We always directly return HPE_PAUSED when we are upgrading. I think probably we can add an ENVOY_BUG here that the error is not HPE_PAUSED_UPGRADE, since ideally we'd like to detect if this is not the case while we're canarying.

antoniovicente commented:
CI is still having problems.

/wait

antoniovicente commented:
coverage seems to be failing, this time due to a flaky test?

//test/extensions/filters/network/kafka/broker/integration_test:kafka_broker_integration_test FAILED in 3 out of 3 in 0.3s

/wait

asraa commented:
Yeah, I"ve been trying to repro it, it was happening before as well. I've tried -c opt, -c dbg with repeated test numbers, I can't repro it so far (and nothing about my change should be affecting it... it's also running with the existing codec in coverage)

asraa commented:
I'm still getting a too many args on my gcc build for //test/exe:version_out_test

moderation commented:
llhttp 6.0.2 is out https://github.com/nodejs/llhttp/releases/tag/release%2Fv6.0.2

indutny commented:
I'd suggest updating to 6.0.2. It fixes a nasty 64bit to 32bit truncation bug in content-length handling, that I tracked down by pure accident!

moderation commented:
@indutny but the nasty bug you found will be tested by tests now right? :wink:

indutny commented:
Of course! I blame not having any 32bit machines at hand for not seeing it earlier.

antoniovicente commented:
@asraa could you please merge upstream/main and update to the latest llhttp version as requested above?

/wait

asraa commented:
Yeah, I can go ahead and do that -- it won't fix the cache filter test issue but i'll push the merge.

antoniovicente commented:
I tried to do some debugging but the conclusion I got to was along the lines that the stall happens even if I disable the cache filter.  This may be a real issue processing multiple requests on the same connection with the new parser.  Is there a chance that some state is not being reset in the parser between requests?

Debugging the state of the llhttp parser isn't easy between the typescript to c transcoding and it being an external dep for the build.  Would it be possible to add methods to dump info about the state of the parser or debug accessors which can be used to implement said debug methods in ParserImpl?

moderation commented:
`llhttp` 6.0.3 is out - https://github.com/nodejs/llhttp/releases/tag/release%2Fv6.0.3

antoniovicente commented:
Please see precheck format presubmit error.

/wait

```
021-06-17T20:30:38.8262526Z ERROR: ./docs/root/version_history/current.rst:82: Version history not in alphabetical order (crash support vs admission control): please check placement of line
2021-06-17T20:30:38.8264348Z  * admission control: added :ref:`admission control <envoy_v3_api_field_extensions.filters.http.admission_control.v3alpha.AdmissionControl.rps_threshold>` option that when average RPS of the sampling window is below this threshold, the filter will not throttle requests. Added :ref:`admission control <envoy_v3_api_field_extensions.filters.http.admission_control.v3alpha.AdmissionControl.max_rejection_probability>` option to set an upper limit on the probability of rejection.. 
```

antoniovicente commented:
cache filter integration test still failing

/wait

moderation commented:
`llhttp` 6.0.4 is out - https://github.com/nodejs/llhttp/releases/tag/release%2Fv6.0.4

yanavlasov commented:
Obsolete. QUICHE H/1 code (Balsa) is currently being integrated.
